apps/emqx_s3/src/emqx_s3_client.erl

```diff
@@ -298,186 +302,45 @@ ensure_aws_credentials(
-            {error, Reason}
+        emqx_s3_client_http:request(Url, Method, Headers, Body, Timeout, HttpClient)
     end.
 
-pick_worker_safe(HttpPool, PoolType) ->
-    try
-        {ok, pick_worker(HttpPool, PoolType)}
-    catch
-        error:badarg ->
-            {error, no_ehttpc_pool}
-    end.
-
-pick_worker(HttpPool, random) ->
-    ehttpc_pool:pick_worker(HttpPool);
-pick_worker(HttpPool, hash) ->
-    ehttpc_pool:pick_worker(HttpPool, self()).
-
--define(IS_BODY_EMPTY(Body), (Body =:= undefined orelse Body =:= <<>>)).
--define(NEEDS_NO_BODY(Method), (Method =:= get orelse Method =:= head orelse Method =:= delete)).
-
-make_request(Method, PathQuery, Headers, Body) when
-    ?IS_BODY_EMPTY(Body) andalso ?NEEDS_NO_BODY(Method)
-->
-    {PathQuery, Headers};
-make_request(_Method, PathQuery, Headers, Body) when ?IS_BODY_EMPTY(Body) ->
-    {PathQuery, [{<<"content-length">>, <<"0">>} | Headers], <<>>};
-make_request(_Method, PathQuery, Headers, Body) ->
-    {PathQuery, Headers, Body}.
-
-format_request({PathQuery, Headers, _Body}) -> {PathQuery, Headers, <<"...">>};
-format_request({PathQuery, Headers}) -> {PathQuery, Headers}.
-
-with_path_and_query_only(Url, Fun) ->
-    case string:split(Url, "//", leading) of
-        [_Scheme, UrlRem] ->
-            case string:split(UrlRem, "/", leading) of
-                [_HostPort, PathQuery] ->
-                    Fun([$/ | PathQuery]);
-                _ ->
-                    {error, {invalid_url, Url}}
-            end;
-        _ ->
-            {error, {invalid_url, Url}}
-    end.
-
-%% We need some header conversions to tie the emqx_s3, erlcloud and ehttpc APIs together.
-
+%% We need some header conversions to tie the emqx_s3, erlcloud and Hackney APIs together.
 %% The request header flow is:
-
-%% UserHeaders -> [emqx_s3_client API] -> ErlcloudRequestHeaders0 ->
-%% -> [erlcloud API] -> ErlcloudRequestHeaders1 -> [emqx_s3_client injected request_fun] ->
-%% -> EhttpcRequestHeaders -> [ehttpc API]
-
+%%   UserHeaders --(emqx_s3_client)-> ErlcloudRequestHeaders
+%%   ErlcloudRequestHeaders --(erlcloud_httpc)--(emqx_s3_client_http)-> HackneyHeaders
+%%   HackneyHeaders --(hackney)-> HTTP Request
+%%
 %% The response header flow is:
-
-%% [ehttpc API] -> EhttpcResponseHeaders -> [emqx_s3_client injected request_fun] ->
-%% -> ErlcloudResponseHeaders0 -> [erlcloud API] -> [emqx_s3_client API]
-
+%%   HTTP Response --(hackney)-> HackneyHeaders
+%%   HackneyHeaders --(emqx_s3_client_http)--(erlcloud_httpc)-> ErlcloudResponseHeders
+%%   ErlcloudResponseHeders --(emqx_s3_client)-> ...
+%%
 %% UserHeders (emqx_s3 API headers) are maps with string/binary keys.
 %% ErlcloudRequestHeaders are lists of tuples with string keys and iodata values
 %% ErlcloudResponseHeders are lists of tuples with lower case string keys and iodata values.
-%% EhttpcHeaders are lists of tuples with binary keys and iodata values.
+%% HackneyHeaders are lists of tuples with binary keys and iodata values.
```

Comment:
Should't this module completely forget about header conversion and the `emqx_s3_client_http` actual backend (Hackney)?